examples/quads.c

```diff
@@ -0,0 +1,216 @@
+#include <string.h>
+#include <strings.h>
+#include <time.h>
+#include <unistd.h>
+#include <wayland-server-core.h>
+#include <wlr/backend.h>
+#include <wlr/backend/session.h>
+#include <wlr/render/wlr_renderer.h>
+#include <wlr/types/wlr_keyboard.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_output_layout.h>
+#include <wlr/types/wlr_input_device.h>
+#include <wlr/types/wlr_matrix.h>
+#include <wlr/util/log.h>
+#include <xkbcommon/xkbcommon.h>
+
+struct sample_state {
+	struct wl_display *display;
+	struct wl_listener new_output;
+	struct wl_listener new_input;
+	struct timespec last_frame;
+	struct wlr_renderer *renderer;
+	struct wl_list outputs;
+};
+
+struct sample_output {
+	struct sample_state *sample;
+	struct wlr_output *output;
+	struct wl_listener frame;
+	struct wl_listener destroy;
+	struct wl_list link;
+};
+
+struct sample_keyboard {
+	struct sample_state *sample;
+	struct wlr_input_device *device;
+	struct wl_listener key;
+	struct wl_listener destroy;
+};
+
+static void output_frame_notify(struct wl_listener *listener, void *data) {
+	struct sample_output *sample_output = wl_container_of(listener, sample_output, frame);
+	struct sample_state *sample = sample_output->sample;
+	struct wlr_output *wlr_output = sample_output->output;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+
+	int32_t width, height;
+	wlr_output_effective_resolution(wlr_output, &width, &height);
+
+	wlr_output_attach_render(wlr_output, NULL);
+	wlr_renderer_begin(sample->renderer, wlr_output->width, wlr_output->height);
+	wlr_renderer_clear(sample->renderer, (float[]){0.25f, 0.25f, 0.25f, 1});
+
+	static float rotation = 0.f;
+	for (int y = -128; y < height; y += 128) {
+		for (int x = -128; x < width; x += 128) {
+			struct wlr_box box = {
+				.x = x,
+				.y = y,
+				.width = 128,
+				.height = 128,
+			};
+
+			float color[] = {
+				255.f / x,
+				255.f / y,
+				255.f / (x + y),
+				1.f
+			};
+
+			float matrix[9];
+			wlr_matrix_project_box(matrix, &box, WL_OUTPUT_TRANSFORM_NORMAL,
+					rotation, wlr_output->transform_matrix);
+
+			wlr_render_quad_with_matrix(sample->renderer, color, matrix);
+		}
+	}
+
+	wlr_renderer_end(sample->renderer);
+	wlr_output_commit(wlr_output);
+
+	//TODO rotate with a delta time
+	rotation += 0.05;
+	if (rotation > 2 * M_PI) {
+		rotation = 0.f;
+	}
+
+	sample->last_frame = now;
+}
+
+static void output_remove_notify(struct wl_listener *listener, void *data) {
+	struct sample_output *sample_output = wl_container_of(listener, sample_output, destroy);
+	wl_list_remove(&sample_output->frame.link);
+	wl_list_remove(&sample_output->destroy.link);
+	free(sample_output);
+}
+
+static void new_output_notify(struct wl_listener *listener, void *data) {
+	struct wlr_output *output = data;
+	struct sample_state *sample = wl_container_of(listener, sample, new_output);
+	struct sample_output *sample_output = calloc(1, sizeof(struct sample_output));
+	if (!wl_list_empty(&output->modes)) {
```

Comment:
Using `wlr_output_preferred_mode` is cleaner here. It sounds like only `simple` was using it, fixed in https://github.com/swaywm/wlroots/commit/fd7e565ce3ea463b725fcd5b4291411f030a4e2c.